compiler/src/iree/compiler/ExternalInterfaces/EncodingExternalModels.cpp

```diff
@@ -106,19 +109,185 @@ struct ContractionOpPropagationInterface
+                // If the source op is the encoding op, we can just add the
+                // source to new operands vector and continue.
+                Operation *sourceOp = operand->get().getDefiningOp();
+                if (sourceOp && sourceOp == encodingOp) {
+                  encodedOperands.push_back(encodingOp.getSource());
+                  continue;
+                }
+
+                auto operandType =
+                    cast<RankedTensorType>(operand->get().getType());
+                auto resType = RankedTensorType::get(
+                    operandType.getShape(), operandType.getElementType(),
+                    encoding);
+                Value encodedInput =
+                    rewriter.create<IREE::Encoding::SetEncodingOp>(
+                        loc, resType, operand->get());
+                result.generatedEncodingOps.push_back(
+                    encodedInput.getDefiningOp());
+                encodedOperands.push_back(encodedInput);
+              }
+
+              SmallVector<Type> resultEncodingTypes;
+              resultEncodingTypes.reserve(resultEncodings.size());
+              for (auto &&[operand, encoding] :
+                   llvm::zip_equal(genericOp.getDpsInits(), resultEncodings)) {
+                auto emptyOp = operand.getDefiningOp<tensor::EmptyOp>();
+                if (!emptyOp) {
+                  return failure();
+                }
+                auto resultEncodingType =
+                    dyn_cast<RankedTensorType>(emptyOp.getResult().getType())
+                        .cloneWithEncoding(encoding);
+
+                // Create encoded generic op.
+                rewriter.setInsertionPointAfter(emptyOp);
+                Value encodedInit = rewriter.create<tensor::EmptyOp>(
+                    loc, emptyOp.getType().getShape(),
+                    resultEncodingType.getElementType(),
+                    emptyOp.getDynamicSizes(), encoding);
+                resultEncodingTypes.push_back(resultEncodingType);
+                encodedOperands.push_back(encodedInit);
+              }
+
+              // Create the generic op with new encoded operands.
+              rewriter.setInsertionPointAfter(genericOp);
+              auto encodedGenericOp = clone(
+                  rewriter, genericOp, resultEncodingTypes, encodedOperands);
+
+              // Create the replacement unset encoding ops.
+              for (unsigned i = 0; i < genericOp->getNumResults(); i++) {
+                auto resultType = cast<RankedTensorType>(
+                                      encodedGenericOp.getResult(i).getType())
+                                      .dropEncoding();
+                auto newUnsetEncoding =
+                    rewriter.create<IREE::Encoding::UnsetEncodingOp>(
+                        encodingOp.getLoc(), resultType,
+                        encodedGenericOp.getResult(i),
+                        encodingOp.getResultDims());
```

Comment:
nit: Iterate over the results directly?
```suggestion
              for (OpResult genericResult : encodedGenericOp->getOpResults()) {
                auto resultType = cast<RankedTensorType>(genericResult.getType())
                                      .dropEncoding();
                auto newUnsetEncoding =
                    rewriter.create<IREE::Encoding::UnsetEncodingOp>(
                        encodingOp.getLoc(), resultType, genericResult,
                        encodingOp.getResultDims());
```